lambdas/select-cri/src/main/java/uk/gov/di/ipv/core/selectcri/SelectCriHandler.java

```diff
@@ -217,8 +246,98 @@ private APIGatewayProxyResponseEvent getJourneyResponse(String criId) {
                 HttpStatus.SC_OK, new JourneyResponse(String.format(CRI_START_JOURNEY, criId)));
     }
 
+    private APIGatewayProxyResponseEvent getJourneyPyiNoMatchResponse() {
+        return ApiGatewayResponseGenerator.proxyJsonResponse(
+                HttpStatus.SC_OK, new JourneyResponse("/journey/pyi-no-match"));
+    }
+
     private boolean userHasNotVisited(
             List<VisitedCredentialIssuerDetailsDto> visitedCredentialIssuers, String criId) {
         return visitedCredentialIssuers.stream().noneMatch(cri -> cri.getCriId().equals(criId));
     }
+
+    private Optional<APIGatewayProxyResponseEvent> getCriResponse(
+            List<VisitedCredentialIssuerDetailsDto> visitedCredentialIssuers,
+            BiPredicate<CredentialEvidenceItem, Gpg45Profile> gpg45Validator,
+            Gpg45Profile gpg45Profile,
+            String criId,
+            String userId)
+            throws ParseException {
+        if (userHasNotVisited(visitedCredentialIssuers, criId)) {
+            return Optional.of(getJourneyResponse(criId));
+        } else {
+            Optional<APIGatewayProxyResponseEvent> failedJourneyResponse =
+                    getFailedJourneyResponse(
+                            visitedCredentialIssuers, gpg45Validator, gpg45Profile, criId, userId);
+
+            if (failedJourneyResponse.isPresent()) {
+                if (criId.equals(dcmawCriId)) {
+                    LOGGER.info("Routing user to web journey");
+                    return Optional.of(getNextWebJourneyCri(visitedCredentialIssuers, userId));
+                }
+                LOGGER.info("Routing user to failed journey path");
+            }
+            return failedJourneyResponse;
+        }
+    }
+
+    private Optional<APIGatewayProxyResponseEvent> getFailedJourneyResponse(
+            List<VisitedCredentialIssuerDetailsDto> visitedCredentialIssuers,
+            BiPredicate<CredentialEvidenceItem, Gpg45Profile> validator,
+            Gpg45Profile gpg45Profile,
+            String criId,
+            String userId)
+            throws ParseException {
+        Optional<VisitedCredentialIssuerDetailsDto> criVisitDetails =
+                visitedCredentialIssuers.stream()
+                        .filter(cri -> cri.getCriId().equals(criId))
+                        .findFirst();
```

Comment:
Due to calling `userHasNotVisited` on line 266 this should always find a `VisitedCredentialIssuerDetailsDto`. So I think you could use `orElseThrow()` and remove the need for an Optional here.
```suggestion
        VisitedCredentialIssuerDetailsDto criVisitDetails =
                visitedCredentialIssuers.stream()
                        .filter(cri -> cri.getCriId().equals(criId))
                        .findFirst().orElseThrow();
```